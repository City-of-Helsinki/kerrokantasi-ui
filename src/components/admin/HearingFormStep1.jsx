/* eslint-disable react/forbid-prop-types */
/* eslint-disable jsx-a11y/no-static-element-interactions */
import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage, useIntl } from 'react-intl';
import { Button, Combobox, TextInput } from 'hds-react';
import { connect, useDispatch } from 'react-redux';

import HearingLanguages from './HearingLanguages';
import MultiLanguageTextField from '../forms/MultiLanguageTextField';
import LabelModal from './LabelModal';
import ContactModal from './ContactModal';
import { contactShape, hearingShape, labelShape, organizationShape } from '../../types';
import getAttr from '../../utils/getAttr';
import Icon from '../../utils/Icon';
import { addLabel, addContact, saveContact, fetchHearingEditorContactPersons } from '../../actions/hearingEditor';
import ContactCard from '../ContactCard';

const HearingFormStep1 = ({
  contactPersons,
  errors,
  hearing,
  hearingLanguages,
  labels: labelOptions,
  language,
  organizations,
  onHearingChange,
  onLanguagesChange,
  onContinue,
}) => {

  const intl = useIntl();
  const dispatch = useDispatch();

  const [showLabelModal, setShowLabelModal] = useState(false);
  const [contactInfo, setContactInfo] = useState({});
  const [showContactModal, setShowContactModal] = useState(false);
  const [selectedLabels, setSelectedLabels] = useState(
    hearing?.labels?.map(({ id }) => id) || []
  );
  const [selectedContacts, setSelectedContacts] = useState(
    hearing?.contact_persons?.filter(Boolean)?.map((person) => person?.id) || []
  );
  
  const onLabelsChange = (labels) => {
    const newLabels = labelOptions.filter((item) => labels.some((label) => item.id === label.id));

    setSelectedLabels(newLabels.map(({ id }) => id));
    onHearingChange(
      'labels',
      newLabels.map(({ id }) => id),
    );
  };

  const onContactsChange = (contacts) => {
    const newContacts = contactPersons.filter((item) => contacts.some((contact) => item.id === contact.id));
    setSelectedContacts(newContacts.filter(Boolean).map(({ id }) => id));
    onHearingChange(
      'contact_persons',
      newContacts.map(({ id }) => id),
    );
  };

  const onCreateLabel = (label) => {
    dispatch(addLabel(label, selectedLabels));
  };

  const onCreateContact = async (contact) => {
    try {
      await dispatch(addContact(contact, selectedContacts));
      await dispatch(fetchHearingEditorContactPersons());
      return true;
    } catch (error) {
      return false;
    }
  };

  const onEditContact = async (contact) => {
    try { 
      await dispatch(saveContact(contact));
      return true;
    } catch (error) {
      return false
    }
  };

  const openLabelModal = () => {
    setShowLabelModal(true);
  };

  const closeLabelModal = () => {
    setShowLabelModal(false);
  };

  const openContactModal = (data) => {
    setShowContactModal(true);
    setContactInfo(data);
  };

  const closeContactModal = () => {
    setShowContactModal(false);
  };

  return (
    <div className='form-step'>
      <HearingLanguages hearingLanguages={hearingLanguages} onChange={onLanguagesChange} />
      <MultiLanguageTextField
        error={errors.title}
        languages={hearingLanguages}
        onBlur={(value) => onHearingChange('title', value)}
        onChange={(value) => onHearingChange('title', value)}
        labelId='title'
        hint={
          <>
            200 <FormattedMessage id='characters' />
          </>
        }
        maxLength={200}
        value={hearing.title}
        name='title'
        placeholderId='titlePlaceholder'
        required
      />
      <div className='hearing-form-row'>
        <div id='hearingLabels' className='hearing-form-column'>
          <Combobox
            multiselect
            name='labels'
            defaultValue={hearing.labels.map((opt) => ({
              id: opt.id,
              title: getAttr(opt.label, language),
              label: getAttr(opt.label, language),
            }))}
            onChange={onLabelsChange}
            options={labelOptions.map((opt) => ({
              id: opt.id,
              title: getAttr(opt.label, language),
              label: getAttr(opt.label, language),
            }))}
            placeholder={intl.formatMessage({ id: 'hearingLabelsPlaceholder' })}
            clearButtonAriaLabel='Poista'
            selectedItemRemoveButtonAriaLabel='Poista {value}'
            toggleButtonAriaLabel='Avaa'
            required
            invalid={!!errors.labels}
            error={errors.labels}
            label={intl.formatMessage({ id: 'hearingLabels' })}
            id='labels'
          />
          <Button
            size='small'
            className='kerrokantasi-btn pull-right action-button'
            onClick={() => openLabelModal()}
            data-testid='add-new-label'
          >
            <Icon className='icon' name='plus' />
          </Button>
        </div>
        <div id='hearingSlug'>
          <TextInput
            id='slug'
            name='slug'
            label={<FormattedMessage id='hearingSlug' />}
            value={hearing.slug}
            placeholder={intl.formatMessage({ id: 'hearingSlugPlaceholder' })}
            onChange={(value) => onHearingChange('slug', value)}
            required
            invalid={!!errors.slug}
            errorText={errors.slug}
          />
        </div>
      </div>
      <div>
        <div id='hearingContacts' className='hearing-form-column'>
          <Combobox
            multiselect
            name='contact_persons'
            onChange={onContactsChange}
            optionKeyField='id'
            value={hearing.contact_persons.filter(Boolean).map((person) => ({
              id: person.id,
              title: person.name,
              label: person.name,
            }))}
            options={contactPersons.map((person) => ({ id: person.id, title: person.name, label: person.name }))}
            placeholder={intl.formatMessage({ id: 'hearingContactsPlaceholder' })}
            label={intl.formatMessage({ id: 'hearingContacts' })}
            required
            helper={<FormattedMessage id='hearingContactsHelpText' />}
            invalid={!!errors.contact_persons}
            error={errors.contact_persons}
            id='contact_persons'
          />
          <Button
            size='small'
            className='kerrokantasi-btn pull-right action-button'
            onClick={() => openContactModal({})}
            data-testid='add-new-contact'
          >
            <Icon className='icon' name='plus' />
          </Button>
        </div>
      </div>

      <ul className='edit-contacts-list hearing-contacts'>
        {selectedContacts &&
          selectedContacts.map((item) => {
            const contact = contactPersons.find((option) => option.id === item);
            if (!contact) return null;
            return (
              <li className='hearing-contact'>
                <ContactCard {...contact} />
                <Button
                  className='kerrokantasi-btn'
                  variant='supplementary'
                  size='small'
                  onClick={() => openContactModal(contact)}
                  aria-label={`Muokkaa yhteyshenkilöä: ${contact.name}`}
                >
                  <Icon className='icon' name='edit' />
                </Button>
              </li>
            );
          })}
      </ul>
      <div className='step-footer'>
        <Button className='kerrokantasi-btn' onClick={onContinue}>
          <FormattedMessage id='hearingFormNext' />
        </Button>
      </div>
      <LabelModal isOpen={showLabelModal} onClose={closeLabelModal} onCreateLabel={onCreateLabel} />
      <ContactModal
        contactInfo={contactInfo}
        isOpen={showContactModal}
        onClose={closeContactModal}
        onCreateContact={onCreateContact}
        onEditContact={onEditContact}
        organizations={organizations}
      />
    </div>
  );
};

HearingFormStep1.propTypes = {
  contactPersons: PropTypes.arrayOf(contactShape),
  errors: PropTypes.object,
  hearing: hearingShape,
  hearingLanguages: PropTypes.arrayOf(PropTypes.string),
  labels: PropTypes.arrayOf(labelShape),
  language: PropTypes.string,
  onContinue: PropTypes.func,
  onHearingChange: PropTypes.func,
  onLanguagesChange: PropTypes.func,
  organizations: PropTypes.arrayOf(organizationShape),
};

HearingFormStep1.contextTypes = {
  language: PropTypes.string,
};

const mapStateToProps = (state) => ({
  language: state.language,
});

export default connect(mapStateToProps, null)(HearingFormStep1);
