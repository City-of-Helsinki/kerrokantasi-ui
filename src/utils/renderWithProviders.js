/* eslint react/prop-types: 0 */
import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { LoginProvider } from 'hds-react';
import { IntlProvider } from 'react-intl';

import createStore from "../createStore";
import { userOidcConfig, apiTokenClientConfig } from './oidcConfig';

const loginProviderConfig = {
  userManagerSettings: userOidcConfig,
  apiTokensClientSettings: apiTokenClientConfig,
  sessionPollerSettings: { pollIntervalInMs: 10000 },
}

const renderWithProviders = (
  ui,
  { preloadedState = {}, locale = 'fi', history, store = createStore(preloadedState), ...renderOptions } = {},
) => {
  const Wrapper = ({ children }) => (
    <LoginProvider {...loginProviderConfig}>
      <IntlProvider locale={locale}>
        <Provider store={store}>{children}</Provider>
      </IntlProvider>
    </LoginProvider>
  );
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};


export default renderWithProviders;
