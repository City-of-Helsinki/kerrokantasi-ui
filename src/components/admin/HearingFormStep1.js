import React from 'react';
import PropTypes from 'prop-types';
import Select from 'react-select';
import {injectIntl, intlShape, FormattedMessage} from 'react-intl';
import map from 'lodash/map';

import Button from 'react-bootstrap/lib/Button';
import Col from 'react-bootstrap/lib/Col';
import ControlLabel from 'react-bootstrap/lib/ControlLabel';
import FormControl from 'react-bootstrap/lib/FormControl';
import FormGroup from 'react-bootstrap/lib/FormGroup';
import InputGroup from 'react-bootstrap/lib/InputGroup';
import Row from 'react-bootstrap/lib/Row';

import HearingLanguages from './HearingLanguages';
import MultiLanguageTextField from '../forms/MultiLanguageTextField';
import LabelModal from './LabelModal';
import ContactModal from './ContactModal';
import {
  contactShape,
  hearingShape,
  labelShape
} from '../../types';
import getAttr from '../../utils/getAttr';
import Icon from '../../utils/Icon';
import {getDocumentOrigin, getValidationState} from '../../utils/hearingEditor';

import {addLabel, addContact, saveContact} from '../../actions/hearingEditor';

class HearingFormStep1 extends React.Component {
  constructor(props) {
    super(props);
    this.onChange = this.onChange.bind(this);
    this.onLabelsChange = this.onLabelsChange.bind(this);
    this.onContactsChange = this.onContactsChange.bind(this);

    this.state = {
      showLabelModal: false,
      contactInfo: {},
      showContactModal: false,
      selectedLabels: map(this.props.hearing.labels, ({id}) => id),
      selectedContacts: map(this.props.hearing.contact_persons, ({id}) => id)
    };
  }

  onChange(event) {
    if (this.props.onHearingChange) {
      // Propagate interesting changes to parent components
      const {name: field, value} = event.target;
      this.props.onHearingChange(field, value);
    }
  }

  onLabelsChange(selectedLabels) {
    this.setState({selectedLabels: selectedLabels.map(({id}) => id)});
    this.props.onHearingChange("labels", selectedLabels.map(({id}) => id));
  }

  onContactsChange(selectedContacts) {
    this.setState({ selectedContacts: selectedContacts.map(({id}) => id) });
    this.props.onHearingChange("contact_persons", selectedContacts.map(({id}) => id));
  }

  onCreateLabel(label) {
    this.props.dispatch(addLabel(label, this.state.selectedLabels));
  }

  onCreateContact(contact) {
    this.props.dispatch(addContact(contact, this.state.selectedContacts));
  }

  onEditContact(contact) {
    this.props.dispatch(saveContact(contact));
  }

  openLabelModal() {
    this.setState({ showLabelModal: true });
  }

  closeLabelModal() {
    this.setState({ showLabelModal: false });
  }

  openContactModal(contactInfo) {
    this.setState({ showContactModal: true, contactInfo });
  }

  closeContactModal() {
    this.setState({ showContactModal: false });
  }

  render() {
    const {
      errors,
      hearing,
      hearingLanguages,
      intl: {formatMessage},
      labels: labelOptions,
      contactPersons: contactOptions,
      onHearingChange,
      onLanguagesChange,
    } = this.props;
    const {language} = this.context;

    return (
      <div className="form-step">

        <HearingLanguages
          hearingLanguages={hearingLanguages}
          onChange={onLanguagesChange}
        />

        <MultiLanguageTextField
          error={errors.title}
          languages={hearingLanguages}
          onBlur={(value) => onHearingChange('title', value)}
          labelId="title"
          maxLength={200}
          value={hearing.title}
          name="title"
          placeholderId="titlePlaceholder"
          required
        />

        <Row>
          <Col md={6}>
            <FormGroup controlId="hearingLabels" validationState={getValidationState(errors, 'labels')}>
              <ControlLabel>
                <FormattedMessage id="hearingLabels">{txt => txt + '*'}</FormattedMessage>
              </ControlLabel>
              <div className="label-elements">
                <Select
                  multi
                  clearAllText="Poista"
                  clearValueText="Poista"
                  name="labels"
                  onChange={this.onLabelsChange}
                  options={labelOptions
                    .map((opt) => ({...opt, title: getAttr(opt.label, language), label: getAttr(opt.label, language)}))
                  }
                  placeholder={formatMessage({id: "hearingLabelsPlaceholder"})}
                  simpleValue={false}
                  value={hearing.labels
                    .map((label) => ({...label, title: 'Poista', label: getAttr(label.label, language)}))
                  }
                  valueKey="frontId"
                  menuContainerStyle={{zIndex: 10}}
                />
                <Button bsStyle="primary" className="pull-right add-label-button" onClick={() => this.openLabelModal()}>
                  <Icon className="icon" name="plus"/>
                </Button>
              </div>
            </FormGroup>
          </Col>
          <Col md={6}>
            <FormGroup controlId="hearingSlug" validationState={getValidationState(errors, 'slug')}>
              <ControlLabel><FormattedMessage id="hearingSlug"/>*</ControlLabel>
              <InputGroup>
                <InputGroup.Addon>{getDocumentOrigin()}</InputGroup.Addon>
                <FormControl
                  type="text"
                  name="slug"
                  defaultValue={hearing.slug}
                  placeholder={formatMessage({id: "hearingSlugPlaceholder"})}
                  onBlur={this.onChange}
                />
              </InputGroup>
            </FormGroup>
          </Col>
        </Row>

        <FormGroup controlId="hearingContacts" validationState={getValidationState(errors, 'contact_persons')}>
          <ControlLabel><FormattedMessage id="hearingContacts"/>*</ControlLabel>
          <div className="contact-elements">
            <Select
              valueRenderer={(options) => (
                <span style={{cursor: 'pointer'}} onMouseDown={() => { this.openContactModal(options); }}>
                  {options.name}
                </span>
              )}
              labelKey="name"
              multi
              name="contacts"
              onChange={this.onContactsChange}
              options={contactOptions}
              placeholder={formatMessage({id: "hearingContactsPlaceholder"})}
              simpleValue={false}
              value={hearing.contact_persons.map(person => ({...person}))}
              valueKey="id"
            />
            <Button
              bsStyle="primary"
              className="pull-right add-contact-button"
              onClick={() => this.openContactModal({})}
            >
              <Icon className="icon" name="plus"/>
            </Button>
          </div>
        </FormGroup>
        <div className="step-footer">
          <Button bsStyle="default" onClick={this.props.onContinue}>
            <FormattedMessage id="hearingFormNext"/>
          </Button>
        </div>
        <LabelModal
          isOpen={this.state.showLabelModal}
          onClose={this.closeLabelModal.bind(this)}
          onCreateLabel={this.onCreateLabel.bind(this)}
        />
        <ContactModal
          contactInfo={this.state.contactInfo}
          isOpen={this.state.showContactModal}
          onClose={this.closeContactModal.bind(this)}
          onCreateContact={this.onCreateContact.bind(this)}
          onEditContact={this.onEditContact.bind(this)}
        />
      </div>
    );
  }
}

HearingFormStep1.propTypes = {
  contactPersons: PropTypes.arrayOf(contactShape),
  dispatch: PropTypes.func,
  errors: PropTypes.object,
  hearing: hearingShape,
  hearingLanguages: PropTypes.arrayOf(PropTypes.string),
  intl: intlShape.isRequired,
  labels: PropTypes.arrayOf(labelShape),
  onContinue: PropTypes.func,
  onHearingChange: PropTypes.func,
  onLanguagesChange: PropTypes.func,
};

HearingFormStep1.contextTypes = {
  language: PropTypes.string
};

const WrappedHearingFormStep1 = injectIntl(HearingFormStep1);

export default WrappedHearingFormStep1;
