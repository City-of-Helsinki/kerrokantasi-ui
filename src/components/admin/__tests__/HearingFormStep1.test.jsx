import React from 'react';
import { MemoryRouter } from 'react-router-dom';

import { UnconnectedHearingFormStep1 } from '../HearingFormStep1';
import renderWithProviders from '../../../utils/renderWithProviders';

const renderComponent = (propOverrides) => {
  const props = {
    hearing: {labels: [], contact_persons: []},
    labels: [],
    contactPersons: [],
    hearingLanguages: [],
    organizations: [],
    errors: {title: ''},
    ...propOverrides,
  };

  return renderWithProviders(
    <MemoryRouter>
      <UnconnectedHearingFormStep1 {...props} />
    </MemoryRouter>,
  );
};

describe('<HearingFormStep1 />', () => {
  it('should render correctly', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });
});
