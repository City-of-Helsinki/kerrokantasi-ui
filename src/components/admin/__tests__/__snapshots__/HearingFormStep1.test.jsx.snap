// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HearingFormStep1 /> should render correctly 1`] = `
<div>
  <div
    class="form-step"
  >
    <fieldset
      class="hearing-languages"
    >
      <legend>
        <span>
          hearingLanguages
        </span>
      </legend>
      <div
        class="hearing-languages__row"
      >
        <span
          class="hearing-languages__language"
        >
          <label
            for="kkEditorLanguageSelector-fi"
          >
            <span>
              inLanguage-fi
            </span>
            <input
              id="kkEditorLanguageSelector-fi"
              name="language"
              type="checkbox"
              value="fi"
            />
          </label>
        </span>
        <span
          class="hearing-languages__language"
        >
          <label
            for="kkEditorLanguageSelector-sv"
          >
            <span>
              inLanguage-sv
            </span>
            <input
              id="kkEditorLanguageSelector-sv"
              name="language"
              type="checkbox"
              value="sv"
            />
          </label>
        </span>
        <span
          class="hearing-languages__language"
        >
          <label
            for="kkEditorLanguageSelector-en"
          >
            <span>
              inLanguage-en
            </span>
            <input
              id="kkEditorLanguageSelector-en"
              name="language"
              type="checkbox"
              value="en"
            />
          </label>
        </span>
      </div>
    </fieldset>
    <fieldset
      class="multi-language-text-field"
    >
      <legend>
        title*
      </legend>
    </fieldset>
    <div
      class="row"
    >
      <div
        class="col-md-6"
      >
        <div
          class="form-group"
        >
          <label
            class="control-label"
            for="hearingLabels"
          >
            hearingLabels*
          </label>
          <div
            class="label-elements"
          >
            <div
              class="Select is-clearable is-searchable Select--multi"
            >
              <div
                class="Select-control"
              >
                <div
                  class="Select-multi-value-wrapper"
                  id="react-select-2--value"
                >
                  <div
                    class="Select-placeholder"
                  >
                    hearingLabelsPlaceholder
                  </div>
                  <div
                    class="Select-input"
                    style="display: inline-block;"
                  >
                    <input
                      aria-activedescendant="react-select-2--value"
                      aria-expanded="false"
                      aria-haspopup="false"
                      aria-owns=""
                      role="combobox"
                      style="box-sizing: content-box; width: 5px;"
                      value=""
                    />
                    <div
                      style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: 100%; font-family: inherit; letter-spacing: normal; text-transform: none;"
                    />
                  </div>
                </div>
                <span
                  class="Select-arrow-zone"
                >
                  <span
                    class="Select-arrow"
                  />
                </span>
              </div>
            </div>
            <button
              class="Button-module_button__1msFE button_hds-button__2A0je Button-module_primary__2LfKB button_hds-button--primary__2NVvO Button-module_size-small__3lizH button_hds-button--small__2NFef kerrokantasi-btn pull-right add-label-button"
              type="button"
            >
              <span
                class="Button-module_label__a4np1 button_hds-button__label__2EQa-"
              >
                <span
                  class="fa fa-plus icon"
                />
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="col-md-6"
      >
        <div
          class="form-group"
        >
          <label
            class="control-label"
            for="hearingSlug"
          >
            <span>
              hearingSlug
            </span>
            *
          </label>
          <span
            class="input-group"
          >
            <span
              class="input-group-addon"
            >
              http://localhost/
            </span>
            <input
              class="form-control"
              id="hearingSlug"
              name="slug"
              placeholder="hearingSlugPlaceholder"
              type="text"
              value=""
            />
          </span>
        </div>
      </div>
    </div>
    <div
      class="form-group"
    >
      <label
        class="control-label"
        for="hearingContacts"
      >
        <span>
          hearingContacts
        </span>
        *
      </label>
      <div
        class="contact-elements"
      >
        <div
          class="Select is-clearable is-searchable Select--multi"
        >
          <div
            class="Select-control"
          >
            <div
              class="Select-multi-value-wrapper"
              id="react-select-3--value"
            >
              <div
                class="Select-placeholder"
              >
                hearingContactsPlaceholder
              </div>
              <div
                class="Select-input"
                style="display: inline-block;"
              >
                <input
                  aria-activedescendant="react-select-3--value"
                  aria-expanded="false"
                  aria-haspopup="false"
                  aria-owns=""
                  role="combobox"
                  style="box-sizing: content-box; width: 5px;"
                  value=""
                />
                <div
                  style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: 100%; font-family: inherit; letter-spacing: normal; text-transform: none;"
                />
              </div>
            </div>
            <span
              class="Select-arrow-zone"
            >
              <span
                class="Select-arrow"
              />
            </span>
          </div>
        </div>
        <button
          class="Button-module_button__1msFE button_hds-button__2A0je Button-module_primary__2LfKB button_hds-button--primary__2NVvO Button-module_size-small__3lizH button_hds-button--small__2NFef kerrokantasi-btn pull-right add-contact-button"
          type="button"
        >
          <span
            class="Button-module_label__a4np1 button_hds-button__label__2EQa-"
          >
            <span
              class="fa fa-plus icon"
            />
          </span>
        </button>
      </div>
      <span
        class="help-block"
      >
        <span>
          hearingContactsHelpText
        </span>
      </span>
    </div>
    <div
      class="step-footer"
    >
      <button
        class="Button-module_button__1msFE button_hds-button__2A0je Button-module_primary__2LfKB button_hds-button--primary__2NVvO kerrokantasi-btn"
        type="button"
      >
        <span
          class="Button-module_label__a4np1 button_hds-button__label__2EQa-"
        >
          <span>
            hearingFormNext
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;
